src/java.base/share/classes/java/lang/SecurityManager.java

```diff
@@ -361,1175 +58,325 @@ public SecurityManager() {
      * the class of that method's caller, and so on.
      *
      * @return  the execution stack.
+     * @see StackWalker
      */
-    protected native Class<?>[] getClassContext();
+    protected Class<?>[] getClassContext() {
+        return StackWalkerHolder.STACK_WALKER
+                .walk(s -> s.map(StackWalker.StackFrame::getDeclaringClass)
+                        .skip(1L)
+                        .toArray(Class[]::new));
+    }
+
+    private static class StackWalkerHolder {
+        static final StackWalker STACK_WALKER =
+            StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE);
```

Comment:
```suggestion
            StackWalker.getInstance(Set.of(Option.DROP_METHOD_INFO, Option.RETAIN_CLASS_REFERENCE));
```

Method info is not needed.

Reply:
Thanks, will fix.

Reply:
SecurityManager::getClassContext hasn't been needed since JDK 9 but we decided to keep the implementation in case there are older versions of logging libraries that extend SecurityManager so they can call this method. What we have not in the jep486 is okay, it would be a bit more efficient if method info is dropped, but I think mostly we want to have any remaining usages of this method to move to StackWalker.

Reply:
We've had logging library maintainers on the core-libs-dev several times in the last 7+ years so I hope there is good awareness of StackWalker. SM.getClassContext is legacy, shouldn't be any reason to use it in 2024.

Reply:
Ok, I'll also add an API note to `getClassContext()` to use `StackWalker` instead.

Reply:
Okay, it already has `@see StackWalker`. My guess is that anything extending SM to call getClassContext is very old code. If that old code is compiled with JDK 17+ then they should get a warning. In any case, we replaced the implementation to use StackWalker so getClassContext will continue to work until the SM class is removed.

Reply:
Fixed in https://github.com/openjdk/jdk/pull/21498/commits/7ea65a6febd18ad8f8fa99cfbb8687e761558594